[[ Bug 22085 ]] Fix use after release of dashes struct lengths field

livecodeali opened the pull request:
This patch fixes an issue that arises when using the dashes offset
opcode of the drawing library - the dashes struct's lengths field
is meant to be re-used for the new dashes with offset, but it is
actually released first, causing the passed-in lengths pointer to be
invalid.

montegoulding commented:
@livecodeali I wonder if this is the correct fix. It seems to me that `MCGContextSetStrokeDashOffset` should just not be calling `MCGContextSetStrokeDashes` with the current dashes lengths pointer. How about:

```
void MCGContextSetStrokeDashOffset(MCGContextRef self, MCGFloat p_offset)
{
    if (!MCGContextIsValid(self))
		return;

    if (self->state->stroke_attr.dashes == nullptr)
    {
            MCGContextSetStrokeDashes(self, p_offset, nullptr, 0);
    }
    else
    {
           self->state->stroke_attr.dashes->phase = p_offset;
    }
}
```
Or implement `MCGDashesSetPhase` if we want the abstraction.

livecodeali commented:
@montegoulding Heh, that's a very good point! Although I'm tempted to keep the change to `MCGContextSetStrokeDashes` as it is more correct anyway

runrevmark commented:
@montegoulding / @livecodeali : You currently have to set either dashes or dash-offset using the SetStrokeDashes call - there is no shortcut. MCGDashes instances are reference counted and immutable - if you directly mutate their phase member then it will change the phase for previously saved frames in the context stack (for example).

livecodeali committed:
[[ Bug 22085 ]] Fix use after release of dashes struct lengths field

This patch fixes an issue that arises when using the dashes offset
opcode of the drawing library - the dashes struct's lengths field
is meant to be re-used for the new dashes with offset, but it is
actually released first, causing the passed-in lengths pointer to be
invalid.

montegoulding commented:
@runrevmark So would the correct thing to do here then be to copy the lengths before the call?

runrevmark commented:
@montegoulding : I don't really see a problem with the proposed patch - you could imagine that MCGDashes *might* have accessors for its internal state; in which case the same ordering of the operations in SetDashes would be required. Copying the dashes beforehand (as SetDashes copies the array itself) seems a little pointless - the three calls MCGContextSetStrokeDashes/DashOffset/DashArray are all manipulating the same state so seems reasonable that they fettle slightly with the internal state of the (essentially private to the MCGContext implementation) MCGDashes object.

montegoulding commented:
OK if you don't see a problem I won't stress it just worried me a bit that the function was clearly not being used as expected when originally implemented, however, I can see that as long as the release is after the copy of the lengths it's a non-issue.

@livecode-vulcan  review ok ff45ed4

livecode-vulcan commented:
💙 review by @montegoulding ok ff45ed4e34c5b42cd67871e120feed4ebebe76a5

runrevmark commented:
@montegoulding Heh - well I suspect the mis-use was my fault - I think I added the separate accessors when I added drawing support in 9. The SetStrokeDashes call is used by MCGraphicsContext (as it always sets stroke styles 'in one go') - in contrast SVG separates dash-array and dash-offset (phase). Looking at the code (all the way through to the usage of Skia path effects), I'm pretty sure phase and dashes are entirely independent so at some point we probably want to ditch MCGDashes / SetStrokeDashes and store the dash-array and dash-offset as separate context members (which would then make this even more of a non-issue!).

livecode-vulcan commented:
😞 test failure ff45ed4e34c5b42cd67871e120feed4ebebe76a5

* try-community-armv6-android-sdk26_ndk16r15: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/548)
* try-community-armv7-android-sdk26_ndk16r15: [failure](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/540)
* try-community-arm64-android-sdk26_ndk16r15: [failure](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/540)
* try-community-x86-android-sdk26_ndk16r15: [pending](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/541)
* try-community-x86_64-android-sdk26_ndk16r15: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/540)
* try-community-js-emscripten-sdk1.35: [failure](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2062)
* try-community-universal-ios-iphoneos12.1: [error](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/290)
* try-community-universal-ios-iphonesimulator12.1: [pending](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/289)
* try-community-universal-mac-macosx10.9: [error](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2094)
* try-community-x86-linux-debian8: [failure](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2084)
* try-community-x86_64-linux-debian8: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2082)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2117)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2054)

livecodepanos commented:
@livecode-vulcan retry ff45ed4

livecode-vulcan commented:
😎 test success ff45ed4e34c5b42cd67871e120feed4ebebe76a5

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/555)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/547)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/547)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/548)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/547)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2069)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/296)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/296)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2100)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2091)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2089)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2124)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2061)
